lib/Optimizer/Transforms/BasisConversion.cpp

```diff
@@ -0,0 +1,155 @@
+
+using namespace mlir;
+
+//===----------------------------------------------------------------------===//
+// Generated logic
+//===----------------------------------------------------------------------===//
+namespace cudaq::opt {
+#define GEN_PASS_DEF_BASISCONVERSIONPASS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+namespace {
+
+struct BasisTarget : public ConversionTarget {
+  struct OperatorInfo {
+    StringRef name;
+    size_t numControls;
+  };
+
+  BasisTarget(MLIRContext &context, ArrayRef<std::string> targetBasis)
+      : ConversionTarget(context) {
+    constexpr size_t unbounded = std::numeric_limits<size_t>::max();
+
+    // Parse the list of target operations and build a set of legal operations
+    for (const std::string &targetInfo : targetBasis) {
+      StringRef option = targetInfo;
+      auto nameEnd = option.find_first_of('(');
+      auto name = option.take_front(nameEnd);
+      if (nameEnd < option.size())
+        option = option.drop_front(nameEnd);
+
+      auto &info = legalOperatorSet.emplace_back(OperatorInfo{name, 0});
+      if (option.consume_front("(")) {
+        option = option.ltrim();
+        if (option.consume_front("n"))
+          info.numControls = unbounded;
+        else
+          option.consumeInteger(10, info.numControls);
+        assert(option.trim().consume_front(")"));
+      }
+    }
+
+    addLegalDialect<func::FuncDialect, arith::ArithDialect,
+                    cudaq::cc::CCDialect>();
+    addDynamicallyLegalDialect<quake::QuakeDialect>([&](Operation *op) {
+      if (auto optor = dyn_cast<quake::OperatorInterface>(op)) {
+        auto name = optor->getName().stripDialect();
+        for (auto info : legalOperatorSet) {
+          if (info.name != name)
+            continue;
+          if (info.numControls == unbounded ||
+              optor.getControls().size() == info.numControls)
+            return info.numControls == optor.getControls().size();
+        }
+        return false;
+      }
+      return true;
+    });
+  }
+
+  SmallVector<OperatorInfo, 8> legalOperatorSet;
+};
+
+//===----------------------------------------------------------------------===//
+// Pass implementation
+//===----------------------------------------------------------------------===//
+
+struct BasisConversion
+    : public cudaq::opt::impl::BasisConversionPassBase<BasisConversion> {
+  using BasisConversionPassBase::BasisConversionPassBase;
+
+  void runOnOperation() override {
+    auto module = getOperation();
+    if (basis.empty()) {
+      module.emitError("Basis conversion requires a target basis");
+      signalPassFailure();
+      return;
+    }
+
+    // First, we walk the whole module in search for controlled `quake.apply`
+    // operations: If present, we conservatively don't do any conversion. We
+    // also collect quantum kernels.
+    //
+    // FIXME: Preventing basis conversion if there is any controlled
```

Comment:
What about adjoint application?

In general, I think it is perfectly fine to have phase orderings in the passes. Before doing a "basis conversion", the compiler must first perform "apply specialization". It would be deeply incorrect for two unrelated passes to attempt to do the same work as the other is already doing. So detecting that a pass that is required hasn't run (or failed) and then the subsequent pass(es) failing is the right way.

On the other hand, it's not clear that anything that Draconian needs to happen in this case. Adding adjoints and controls should be fairly agnostic to the process of replacing gates with more (a lot more) gates. It means certain control gate patterns wouldn't be exposed to this pass, but it seems like a pipeline of "basis conversion 1", "apply specialization", "basis conversion 2" should get to an equivalent place as "apply specialization", "basis conversion". Indeed, that seems like it could be a useful test setup.

Reply:
> What about adjoint application?

Adjoints shouldn't be problematic from the point of view of this pass. 

> Adding adjoints and controls should be fairly agnostic to the process of replacing gates with more (a lot more) gates.

Not true in the case of controls. Decomposition patterns _might_ add sequence of gates that are equivalent to the original gate up to a global phase, which become relative phase when controls are attached to it. 

For example, suppose we have the following program and we decide to target the `phased_rx, cz` basis:
```mlir
// Unitary: | 1  0 | 
//          | 0  i |
func.func @special_r1(%t: !quake.qref) {
  %0 = arith.constant 1.57079632679489660 : f64
  quake.r1 (%0) %t : (f64, !quake.qref) -> ()
  return
}

func.func @kernel(%c: !quake.qref, %t: !quake.qref) {
  quake.apply @special_r1 [%c] %t : (!quake.qref, !quake.qref) -> ()
  return
}
```

After decomposition, the new `@special_r1` will be equivalent to the original one up to a global phase:
```mlir
//           global phase
//          ________^__________
// Unirary: (0.707107,0.707107) * | 1  0 |
//                                | 0  i | 

func.func @special_r1(%t: !quake.qref) {
   %cst = arith.constant -1.5707963267948966 : f64
    %cst_0 = arith.constant 1.5707963267948966 : f64
    %cst_1 = arith.constant 0.000000e+00 : f64
    quake.phased_rx (%cst_0, %cst_1) %arg0 : (f64, f64, !quake.qref) -> ()
    quake.phased_rx (%cst, %cst_0) %arg0 : (f64, f64, !quake.qref) -> ()
    quake.phased_rx (%cst, %cst_1) %arg0 : (f64, f64, !quake.qref) -> ()
    return
}
```

If we then specialize the `quake.apply`, the unitary for `@kernel` will be:
```
// Unirary: | 1  0  0  0  |
//          | 0  1  0  0  |    global phase that became relative phase
//          | 0  0  x  0  | ______________^________________
//          | 0  0  0  xi | (where, x = (0.707107,0.707107))
```

Which is wrong, because it should be:
```
// Unitary: | 1  0  0  0  |
//          | 0  1  0  0  |
//          | 0  0  1  0  | 
//          | 0  0  0  i  | 
```
(or some other unitary that is equivalent up to a global phase)

Having said that, I think `FIXME` was poor wording from me. It just happens that some decomposition patterns are exact, and thus wouldn't suffer from the problem described above. It is an open question whether we might want to apply those patterns while keeping controlled `quake.apply` untouched. So this comment is just a reminder in case we ever do want.